### PR TITLE
[#9420] Consolidate Join RESTFul API

### PR DESCRIPTION
**Pull request by ChooJeremy**
Part of #9420 

**Outline of Solution**

* Nothing much to say here. Changed Join to use DTO, and merged the send emails API together (their code was basically identical.
* Tests have been updated, more checks and balances. Interestingly, the try catch blocks previously didn't even enter their catch blocks because no exception was thrown, and the tests still passed because the throwing wasn't required.
* Frontend API updated, will have merge conflicts with other course.service.ts as well.

Ready for review.

### CODE REVIEW

File: src/test/java/teammates/test/cases/webapi/SendReminderEmailActionTest.java [hunk 1]
```diff
@@ -70,25 +75,25 @@ protected void testExecute() throws Exception {
...
-        remindAction = getAction(submissionParams);
-        r = getJsonResult(remindAction);
+        sendReminderEmailAction = getAction(submissionParams);
+        result = getJsonResult(sendReminderEmailAction);
 
-        msg = (MessageOutput) r.getOutput();
+        msg = (MessageOutput) result.getOutput();
         assertEquals("An email has been sent to " + student1InCourse1.email, msg.getMessage());
 
         ______TS("Typical case: Send email to remind a student to register for the course from student list page");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/SendReminderEmailActionTest.java]: I don't think you mention where is the request come from here.

Check LINE70 also.

File: src/test/java/teammates/test/cases/webapi/SendReminderEmailActionTest.java [hunk 2]
```diff
@@ -143,56 +148,50 @@ protected void testExecute() throws Exception {
...
                 Const.ParamsNames.INSTRUCTOR_EMAIL, invalidEmail,
         };
 
-        try {
-            getAction(addUserIdToParams(instructorId, submissionParams));
-        } catch (EntityNotFoundException e) {
-            assertEquals("Instructor with email " + invalidEmail + " does not exist "
-                    + "in course " + courseId + "!", e.getMessage());
-        }
+        EntityNotFoundException entityNotFoundException = assertThrows(EntityNotFoundException.class, () ->
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/SendReminderEmailActionTest.java]: 👍

File: src/web/services/course.service.ts [hunk 3]
```diff
@@ -24,4 +24,36 @@ export class CourseService {
     };
     return this.httpRequestService.get('/join', paramMap);
   }
+
+  /**
+   * Send reminder emails to everyone.
```
Comment (xpdavid) [inline src/web/services/course.service.ts]: Not to everyone but only to unregistered students. You should change the method name to a proper name also.

**Comment by ChooJeremy** [comment]
Ready for review.

File: src/web/services/course.service.ts [hunk 4]
```diff
@@ -49,7 +49,7 @@ export class CourseService {
   /**
    * Send reminder emails to an instructor.
    */
-  sendReminderEmailToInstructor(courseId: string, instructorEmail: string): Observable<MessageOutput> {
+  remindInstructor(courseId: string, instructorEmail: string): Observable<MessageOutput> {
```
Comment (xpdavid) [inline src/web/services/course.service.ts]: remindInstructorForJoin?

File: src/web/services/course.service.ts [hunk 5]
```diff
@@ -38,7 +38,7 @@ export class CourseService {
   /**
    * Send reminder emails to a student.
    */
-  sendReminderEmailToStudent(courseId: string, studentEmail: string): Observable<MessageOutput> {
+  remindStudent(courseId: string, studentEmail: string): Observable<MessageOutput> {
```
Comment (xpdavid) [inline src/web/services/course.service.ts]: remindStudentForJoin?

File: src/web/services/course.service.ts [hunk 6]
```diff
@@ -26,9 +26,9 @@ export class CourseService {
   }
 
   /**
-   * Send reminder emails to everyone.
+   * Send reminder emails to unregistered students.
    */
-  sendReminderEmail(courseId: string): Observable<MessageOutput> {
+  remindUnregisteredStudents(courseId: string): Observable<MessageOutput> {
```
Comment (xpdavid) [inline src/web/services/course.service.ts]: remindUnregisteredStudentsForJoin?

File: src/web/services/course.service.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  joinCourse(regKey: string, entityType: string): Observable<JoinStatus> {
+    const paramMap: { [key: string]: string } = {
+      key: regKey,
+      entitytype: entityType,
+    };
+    return this.httpRequestService.get('/join', paramMap);
+  }
+
+  /**
+   * Send reminder emails to unregistered students.
```
Comment (xpdavid) [inline src/web/services/course.service.ts]: `Send join reminder emails to unregistered students.`.

Same for other method.

File: src/main/java/teammates/ui/webapi/action/ActionFactory.java [hunk 8]
```diff
@@ -79,6 +79,7 @@
         map(ResourceURIs.LOCAL_DATE_TIME, GET, GetLocalDateTimeInfoAction.class);
         map(ResourceURIs.JOIN, GET, GetCourseJoinStatusAction.class);
         map(ResourceURIs.JOIN, PUT, JoinCourseAction.class);
+        map(ResourceURIs.JOIN_REMIND, POST, SendReminderEmailAction.class);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/ActionFactory.java]: `SendReminderEmailAction` -> `SendJoinReminderEmailAction`.

Let's make it clear as reminder email can be multiple things in our application.

**Comment by ChooJeremy** [comment]
Ready for review.

**Pull request merged**